src/solver/opamCudf.ml

```diff
@@ -1037,19 +1028,120 @@ let extract_explanations packages cudfnv2opam reasons : explanation list =
                   Printf.sprintf "%s: no longer available"
                     (OpamPackage.to_string (cudf2opam p))
                 in
-                `Missing (Some csp, msg, OpamFormula.Empty)
+                let csp = cst_to_string csp in
+                `NoLongerAvailable (csp, msg, OpamFormula.Empty)
               else
               let fdeps = formula_of_vpkgl cudfnv2opam packages deps in
-              let sdeps = OpamFormula.to_string fdeps in
-              `Missing (Some csp, sdeps, fdeps)
+              `Missing (csp, fdeps)
             in
-            if List.mem msg explanations then raise Not_found else
-              msg :: explanations, ct_chains
+            msg :: explanations
           | Dependency _ ->
-            explanations, ct_chains
+            explanations
         with Not_found ->
-          explanations, ct_chains)
-      ([], ct_chains) reasons
+          explanations)
+      [] reasons
+  in
+
+  let rec simplify_formula formula1 formula2 =
+    let open OpamStd.Option.Op in
+    match formula1, formula2 with
+    | Empty, Empty -> Some Empty
+    | Atom (name1, vformula1), Atom (name2, vformula2) ->
+      if OpamPackage.Name.equal name1 name2 then
+        OpamFormula.simplify_version_formula (Or (vformula1, vformula2))
+        >>= fun vformula -> Some (Atom (name1, vformula))
+      else
+        None
+    | Block x, y | x, Block y -> simplify_formula x y
+    | And (x1, x2), And (y1, y2) ->
+        simplify_formula x1 y1 >>= fun z1 ->
+        simplify_formula x2 y2 >>= fun z2 ->
+        Some (And (z1, z2))
+    | Or (x1, x2), Or (y1, y2) ->
+        simplify_formula x1 y1 >>= fun z1 ->
+        simplify_formula x2 y2 >>= fun z2 ->
+        Some (Or (z1, z2))
+    | Empty, _
+    | Atom _, _
+    | And _, _
+    | Or _, _ -> None
+  in
+
+  let explanations =
+    let simplify_cst f f' =
+      match f, f' with
+      | `Invariant, `Invariant
+      | `Request, `Request -> f
+      | `Formula (bold, f), `Formula (bold', f') ->
+        if Bool.equal bold bold' then
+          match simplify_formula f f' with
+          | Some simplified -> `Formula (bold, simplified)
+          | None -> raise (Invalid_argument "")
+        else
+          raise (Invalid_argument "")
+      | `Invariant, _ | `Request, _ | `Formula _, _ ->
+        raise (Invalid_argument "")
```

Comment:
If these case should not happen, it should be an internal error with a message. If it is sure that it won't happen, it should be then be `assert false` with a comment.

Reply:
For the `bold <> bold'` case, if we are sure that it won't happen, it is possible to take directly the first `bold`.

Reply:
The use of `Invalid_argument` here is used because this function is called using `List.fold_left2` and these cases are valid but we want to raise early if the two conflict do not match